### PR TITLE
Make sure we look into /sbin and /usr/sbin for sysctl

### LINKED ISSUE #13138: spack test lists warnings about sysctl on macOS
### Steps to reproduce the issue

```console
$ spack test cmd/compiler_command.py
==================================================================== test session starts ====================================================================
platform darwin -- Python 3.7.3, pytest-3.2.5, py-1.4.34, pluggy-0.4.0
rootdir: /Users/Adam/spack/lib/spack/spack/test, inifile: pytest.ini
collected 2101 items                                                                                                                                         

cmd/compiler_command.py .

================================================================= slowest 20 test durations =================================================================
0.34s call     cmd/compiler_command.py::test_compiler_find_without_paths
0.11s setup    cmd/compiler_command.py::test_compiler_find_without_paths
0.01s teardown cmd/compiler_command.py::test_compiler_find_without_paths
===================================================================== warnings summary ======================================================================
cmd/compiler_command.py::test_compiler_find_without_paths
  /Users/Adam/spack/lib/spack/llnl/util/cpu/detect.py:127: UserWarning: [Errno 2] No such file or directory: 'sysctl': 'sysctl'
    warnings.warn(str(e))
  /Users/Adam/spack/lib/spack/llnl/util/cpu/detect.py:127: UserWarning: [Errno 2] No such file or directory: 'sysctl': 'sysctl'
    warnings.warn(str(e))

-- Docs: http://doc.pytest.org/en/latest/warnings.html
=================================================================== 2100 tests deselected ===================================================================
=================================================== 1 passed, 2100 deselected, 2 warnings in 3.60 seconds ===================================================
```

### Information on your system

- macOS 10.15 Catalina

@alalazo

### PR DESCRIPTION
**Pull request by alalazo**
fixes #13138
closes #13354

This PR ensures that on Darwin we always append `/sbin` and `/usr/sbin` to `PATH`, if they are not already present, when looking for `sysctl`.

### CODE REVIEW

File: lib/spack/llnl/util/cpu/detect.py [hunk 1]
```diff
@@ -75,30 +76,45 @@ def proc_cpuinfo():
...
 def sysctl():
     """Returns a raw info dictionary parsing the output of sysctl."""
+    # Make sure that /sbin and /usr/sbin are in PATH as sysctl is
+    # usually found there
+    child_environment = dict(os.environ.items())
+    search_paths = child_environment.get('PATH', '').split(os.pathsep)
+    for additional_path in ('/sbin', '/usr/sbin'):
+        if additional_path not in search_paths:
+            search_paths.append(additional_path)
+    child_environment['PATH'] = os.pathsep.join(search_paths)
```
Comment (tgamblin) [inline lib/spack/llnl/util/cpu/detect.py]: Suggested refactor: call this function `sysctl_info_dict()` or something more descriptive, then make an internal wrapper around `check_output()` called `sysctl()`:

```python
@info_dict(operating_system='Darwin')
def sysctl_info_dict():
    child_environment = dict(os.environ.items())
    search_paths = child_environment.get('PATH', '').split(os.pathsep)
    for additional_path in ('/sbin', '/usr/sbin'):
        if additional_path not in search_paths:
            search_paths.append(additional_path)
    child_environment['PATH'] = os.pathsep.join(search_paths)

    def sysctl(*args):
        return check_output(['sysctl'] + args, env=child_environment)

    ...
```

Comment (alalazo) [reply lib/spack/llnl/util/cpu/detect.py]: Seems indeed a lot better. I'll probably try to see if turning `info` into a dict literal improves readability.

Comment (alalazo) [reply lib/spack/llnl/util/cpu/detect.py]: Done

**Comment by alalazo** [comment]
Can anybody test drive the latest modifications on Mac?

**Comment by adamjstewart** [comment]
Tested on macOS 10.15:
```
$ spack test cmd/compiler_command.py
========================================================================== test session starts ===========================================================================
platform darwin -- Python 3.7.3, pytest-3.2.5, py-1.4.34, pluggy-0.4.0
rootdir: /Users/Adam/spack/lib/spack/spack/test, inifile: pytest.ini
collected 2164 items                                                                                                                                                      

cmd/compiler_command.py F
======================================================================== short test summary info =========================================================================
FAIL cmd/compiler_command.py::test_compiler_find_without_paths

======================================================================= slowest 20 test durations ========================================================================
0.05s call     cmd/compiler_command.py::test_compiler_find_without_paths
0.03s setup    cmd/compiler_command.py::test_compiler_find_without_paths
0.01s teardown cmd/compiler_command.py::test_compiler_find_without_paths
================================================================================ FAILURES ================================================================================
____________________________________________________________________ test_compiler_find_without_paths ____________________________________________________________________

no_compilers_yaml = None, working_env = None
tmpdir = local('/private/var/folders/21/hwq39zyj4g36x6zjfyl5l8080000gn/T/pytest-of-Adam/pytest-1/test_compiler_find_without_pat0')

    @pytest.mark.regression('11678,13138')
    @pytest.mark.filterwarnings('error')
    def test_compiler_find_without_paths(no_compilers_yaml, working_env, tmpdir):
        with tmpdir.as_cwd():
            with open('gcc', 'w') as f:
                f.write("""\
    #!/bin/bash
    echo "0.0.0"
    """)
            os.chmod('gcc', 0o700)
    
        os.environ['PATH'] = str(tmpdir)
>       output = compiler('find', '--scope=site')

cmd/compiler_command.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../main.py:511: in __call__
    with log_output(out):
../../llnl/util/tty/log.py:289: in __enter__
    self.parent, self.child = multiprocessing.Pipe()
/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/multiprocessing/context.py:61: in Pipe
    from .connection import Pipe
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:967: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:668: in _load_unlocked
    ???
<frozen importlib._bootstrap>:638: in _load_backward_compatible
    ???
../../external/_pytest/assertion/rewrite.py:212: in load_module
    py.builtin.exec_(co, mod.__dict__)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    __all__ = [ 'Client', 'Listener', 'Pipe', 'wait' ]
    
    import io
    import os
    import sys
    import socket
    import struct
    import time
    import tempfile
    import itertools
    
    import _multiprocessing
    
>   from . import util
E   ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__

/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/multiprocessing/connection.py:23: ImportWarning
========================================================================= 2163 tests deselected ==========================================================================
=============================================================== 1 failed, 2163 deselected in 3.71 seconds ================================================================
```

**Comment by alalazo** [comment]
@adamjstewart Ugh, is arch detection still working? Also, can you try adding:
```python
@pytest.mark.filterwarning('ignore:ImportWarning')
```
on top of the test and see if it solves the issue?

**Comment by adamjstewart** [comment]
Arch detection is still working:
```
$ spack arch
darwin-catalina-ivybridge
```
When I try adding your `filterwarning` it still exits with the same error. When I remove both `filterwarning` and `filterwarnings`, it completes successfully.

**Comment by tgamblin** [comment]
I tested this with @alalazo's latest change and it works on my mac.

**Pull request merged**